packages/eslint-plugin/src/rules/thenable-in-promise-aggregators.ts

```diff
@@ -0,0 +1,258 @@
+        const checker = program.getTypeChecker();
+
+        for (const baseType of checker.getBaseTypes(type as ts.InterfaceType)) {
+          if (isPromiseLike(program, baseType)) {
+            return true;
+          }
+        }
+      }
+
+      return false;
+    }
+
+    // Is like PromiseConstructor, or a class/interface extending from Promise
+    function isPromiseLike(program: ts.Program, type: ts.Type): boolean {
+      if (type.isIntersection()) {
+        return type.types.some(t => isPromiseLike(program, t));
+      }
+      if (type.isUnion()) {
+        return type.types.every(t => isPromiseLike(program, t));
+      }
+
+      const symbol = type.getSymbol();
+      if (!symbol) {
+        return false;
+      }
+
+      if (
+        symbol.getName() === 'PromiseConstructor' &&
+        isSymbolFromDefaultLibrary(program, symbol)
+      ) {
+        return true;
+      }
+
+      if (symbol.flags & (ts.SymbolFlags.Class | ts.SymbolFlags.Interface)) {
+        const checker = program.getTypeChecker();
+
+        for (const baseType of checker.getBaseTypes(type as ts.InterfaceType)) {
+          if (isPromiseClassLike(program, baseType)) {
+            return true;
+          }
+        }
+      }
+
+      return false;
+    }
+
+    return {
+      CallExpression(node: TSESTree.CallExpression): void {
+        const callee = skipChainExpression(node.callee);
+        if (callee.type !== AST_NODE_TYPES.MemberExpression) {
+          return;
+        }
+
+        if (callee.computed) {
+          if (
+            callee.property.type !== AST_NODE_TYPES.Literal ||
+            typeof callee.property.value !== 'string' ||
+            !aggregateFunctionNames.includes(callee.property.value)
+          ) {
+            return;
+          }
+        } else if (!aggregateFunctionNames.includes(callee.property.name)) {
+          return;
+        }
+
+        const callerType = services.getTypeAtLocation(callee.object);
+        if (!isPromiseLike(services.program, callerType)) {
+          return;
+        }
+
+        const args = node.arguments;
+        if (args.length < 1) {
+          return;
+        }
+
+        const arg = args[0];
+        if (arg.type === AST_NODE_TYPES.ArrayExpression) {
+          const { elements } = arg;
+          if (elements.length === 0) {
+            return;
+          }
+
+          for (const element of elements) {
+            if (element == null) {
+              context.report({
+                messageId: 'inArray',
+                node: arg,
+              });
+              return;
+            }
+            const elementType = services.getTypeAtLocation(element);
+            if (isTypeAnyType(elementType) || isTypeUnknownType(elementType)) {
+              continue;
+            }
+
+            const originalNode = services.esTreeNodeToTSNodeMap.get(element);
+            if (tsutils.isThenableType(checker, originalNode, elementType)) {
+              continue;
+            }
```

Comment:
this logic (if `isTypeAnyType(sometype) || isTypeUnknownType(sometype)` or `tsutils.isThenableType(checker, originalNode, sometype)`) is used three times in this rule. Maybe we should move it to the function to avoid duplicated code?

Reply:
Hmm; I refactored around a bit, though I still have 3 calls like this. However, I don't think something like this would add anything besides some mental overhead:

```ts
function isTypeAnyOrUnknownType(type: ts.Type): boolean {
  return isTypeAnyType(type) || isTypeUnknownType(type);
}
```

For this to be useful from my current perspective there would need to be a more salient name for the method that is a more useful concept than `checking for any or unknown` and makes the implementation obvious (since it's so short). I looked elsewhere around the codebase and any types are not universally treated the same as unknown types, e.g., I could imagine a flag on this check later, perhaps two flags (one for any and one for unknown), and depending on exactly how those flags work it might make sense to break it out then, but more likely even if you made those flags it would still not be helpful to have a function like this given you would want different errors each which explanations of the flag to use to disable that feature, ect.